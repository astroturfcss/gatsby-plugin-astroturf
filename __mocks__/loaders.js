export default {
  postcss: () => ({
    loader: 'postcss-loader',
    options: {
      ident: 'postcss',
    },
  }),
}
