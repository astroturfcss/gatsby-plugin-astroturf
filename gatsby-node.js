// match with Gatsby's CSS Modules rule
const EXTENSION = '.module.css'

const findJsRule = ({ config, rules }) =>
  config.module.rules.find(({ test }) => test.source === rules.js().test.source)

exports.onCreateWebpackConfig = (
  { actions, getConfig, rules },
  pluginOptions
) => {
  const config = getConfig()

  const jsRule = findJsRule({ config, rules })
  const astroturfJsLoader = {
    loader: 'astroturf/loader',
    options: Object.assign({}, pluginOptions, {
      extension: EXTENSION,
    }),
  }
  jsRule.use.push(astroturfJsLoader)

  actions.replaceWebpackConfig(config)
}
