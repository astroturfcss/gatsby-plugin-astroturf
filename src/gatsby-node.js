// inspired by gatasby-plugin-postcss
// https://github.com/gatsbyjs/gatsby/blob/afd15c262b7c3fd3a39e962a29edefe1974a7380/packages/gatsby-plugin-postcss/src/gatsby-node.js

const EXTENSION = '.astroturf.css'

const findJsRule = ({ config, rules }) =>
  config.module.rules.find(({ test }) => test.source === rules.js().test.source)

const isCssRule = ({ test }) => test.source.endsWith('\\.css$')
const findCssRules = ({ config }) =>
  config.module.rules.find(
    ({ oneOf }) => Array.isArray(oneOf) && oneOf.every(isCssRule)
  )

export const onCreateWebpackConfig = (
  { actions, getConfig, rules, loaders },
  pluginOptions = {}
) => {
  const {
    // eslint-disable-next-line no-unused-vars
    postcssPlugins = [require('postcss-nested')],
    ...options
  } = pluginOptions

  const config = getConfig()

  const jsRule = findJsRule({ config, rules })
  const astroturfJsLoader = {
    loader: 'astroturf/loader',
    options: {
      ...options,
      extension: EXTENSION,
    },
  }
  jsRule.use.push(astroturfJsLoader)

  const cssRules = findCssRules({ config })
  const astroturfCssRule = rules.cssModules()
  astroturfCssRule.test = new RegExp(EXTENSION.replace(/\./g, '\\.') + '$')
  let postcssLoader = astroturfCssRule.use.find(({ loader }) =>
    /\bpostcss-loader\b/.test(loader)
  )
  if (postcssLoader == null) {
    postcssLoader = loaders.postcss()
    astroturfCssRule.use.push(postcssLoader)
    const cssLoader = astroturfCssRule.use.find(({ loader }) =>
      /\bcss-loader\b/.test(loader)
    )
    cssLoader.options.importLoaders = 1
  }
  postcssLoader.options.ident = 'postcss-astroturf'
  postcssLoader.options.plugins = postcssPlugins

  if (cssRules != null) {
    cssRules.oneOf.unshift(astroturfCssRule)
    actions.replaceWebpackConfig(config)
  } else {
    actions.setWebpackConfig({
      module: {
        rules: [astroturfCssRule],
      },
    })
  }
}
