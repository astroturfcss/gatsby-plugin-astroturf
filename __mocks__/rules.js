import loaders from './loaders'

export default {
  js: () => ({
    test: /\.jsx?/,
    use: [
      {
        loader: 'babel-loader',
      },
    ],
  }),
  css: () => ({
    test: /\.css$/,
    use: [
      {
        loader: 'style-loader',
        options: {},
      },
      {
        loader: 'css-loader',
        options: {},
      },
      loaders.postcss(),
    ],
  }),
  cssWithoutPostcss: () => ({
    test: /\.css$/,
    use: [
      {
        loader: 'style-loader',
        options: {},
      },
      {
        loader: 'css-loader',
        options: {},
      },
    ],
  }),
  cssModules: () => ({
    test: /\.module\.css$/,
    use: [
      {
        loader: 'style-loader',
        options: {},
      },
      {
        loader: 'css-loader',
        options: {
          modules: true,
        },
      },
      loaders.postcss(),
    ],
  }),
  cssModulesWithoutPostcss: () => ({
    test: /\.module\.css$/,
    use: [
      {
        loader: 'style-loader',
        options: {},
      },
      {
        loader: 'css-loader',
        options: {
          modules: true,
        },
      },
    ],
  }),
}
