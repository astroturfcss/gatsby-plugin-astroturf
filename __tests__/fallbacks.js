import { onCreateWebpackConfig } from '../src/gatsby-node'
import getConfigWithoutPostcssLoader from '../__mocks__/webpack-config-without-postcss-loader'
import getConfigWithoutCssRules from '../__mocks__/webpack-config-without-css-rules'
import rules from '../__mocks__/rules'
import loaders from '../__mocks__/loaders'

test("when CSS Modules rule doesn't include postcss-loader", () => {
  const actions = { replaceWebpackConfig: jest.fn() }
  onCreateWebpackConfig({
    actions,
    getConfig: getConfigWithoutPostcssLoader,
    rules,
    loaders,
  })
  expect(actions.replaceWebpackConfig).toMatchSnapshot()
})

test('when there are no CSS rules in the configuration', () => {
  const actions = { setWebpackConfig: jest.fn() }
  onCreateWebpackConfig({
    actions,
    getConfig: getConfigWithoutCssRules,
    rules,
    loaders,
  })
  expect(actions.setWebpackConfig).toMatchSnapshot()
})
