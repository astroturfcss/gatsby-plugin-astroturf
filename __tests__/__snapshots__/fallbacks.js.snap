// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when CSS Modules rule doesn't include postcss-loader 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "test": /\\\\\\.jsx\\?/,
              "use": Array [
                Object {
                  "loader": "babel-loader",
                },
                Object {
                  "loader": "astroturf/loader",
                  "options": Object {
                    "extension": ".astroturf.css",
                  },
                },
              ],
            },
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.astroturf\\\\\\.css\\$/,
                  "use": Array [
                    Object {
                      "loader": "style-loader",
                      "options": Object {},
                    },
                    Object {
                      "loader": "css-loader",
                      "options": Object {
                        "modules": true,
                      },
                    },
                    Object {
                      "loader": "postcss-loader",
                      "options": Object {
                        "ident": "postcss-astroturf",
                        "plugins": Array [
                          [Function],
                        ],
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.module\\\\\\.css\\$/,
                  "use": Array [
                    Object {
                      "loader": "style-loader",
                      "options": Object {},
                    },
                    Object {
                      "loader": "css-loader",
                      "options": Object {
                        "modules": true,
                      },
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.css\\$/,
                  "use": Array [
                    Object {
                      "loader": "style-loader",
                      "options": Object {},
                    },
                    Object {
                      "loader": "css-loader",
                      "options": Object {},
                    },
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`when there are no CSS rules in the configuration 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "test": /\\\\\\.astroturf\\\\\\.css\\$/,
              "use": Array [
                Object {
                  "loader": "style-loader",
                  "options": Object {},
                },
                Object {
                  "loader": "css-loader",
                  "options": Object {
                    "modules": true,
                  },
                },
                Object {
                  "loader": "postcss-loader",
                  "options": Object {
                    "ident": "postcss-astroturf",
                    "plugins": Array [
                      [Function],
                    ],
                  },
                },
              ],
            },
          ],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
